platforms/core-runtime/daemon-protocol/src/main/java/org/gradle/launcher/daemon/context/DaemonCompatibilitySpec.java

```diff
@@ -66,7 +67,17 @@ private boolean daemonOptsMatch(DaemonContext potentialContext) {
     private boolean jvmCompatible(DaemonContext potentialContext) {
         DaemonJvmCriteria criteria = desiredContext.getJvmCriteria();
         if (criteria instanceof DaemonJvmCriteria.Spec) {
-            return ((DaemonJvmCriteria.Spec) criteria).isCompatibleWith(potentialContext.getJavaVersion(), potentialContext.getJavaVendor());
+            DaemonJvmCriteria.Spec daemonJvmCriteria = (DaemonJvmCriteria.Spec) criteria;
+            if (daemonJvmCriteria.isCompatibleWith(potentialContext.getJavaVersion(), potentialContext.getJavaVendor())) {
+                if (daemonJvmCriteria.getNativeImageCapable()) {
+                    // Need to assess that $JAVA_HOME/bin/native-image exists
+                    return new File(new File(potentialContext.getJavaHome(), "bin"), OperatingSystem.current().getExecutableName("native-image")).exists();
```

Comment:
💭 Perhaps not now, but it would be nice to re-use `JvmInstallationMetadata` here instead of duplicating this logic.

Reply:
Yes, it would be more involved, we could serialize the `JvmInstallationMetadata` on the daemon (we have code to get it from running JVM in `DefaultJvmMetadataDetector` already) and send it with the context. But it would clean up the context interface as well, since it would only need the one getter for metadata rather than the three it currently has for the same information.

I don't think it should be done in this PR, but it would be a good follow-up.